app/ide-desktop/lib/dashboard/src/services/Backend.ts

```diff
@@ -959,6 +968,8 @@ export interface UpdateProjectRequestBody {
 /** HTTP request body for the "open project" endpoint. */
 export interface OpenProjectRequestBody {
   readonly executeAsync: boolean
+  /** MUST be present on Remote backend; NOT REQUIRED on Local backend. */
```

Comment:
This is still wild to me why do we put these field into `Backend` class, thus violating SOLID principles.

I'm sure, if a field or a method is only suitable for RemoteBackend, it should be described only there.

Reply:
actually it doesn't, this *is* the dependency inversion pattern. all frontend code is *supposed* to be backend agnostic, so that as much as possible of the logic is shared. and so `cognitoCredentials` MUST be provided when possible, because it is *wrong* for the frontend code to know whether it is interacting with a `RemoteBackend`.

Reply:
(if you have a better solution though, i'm happy to accept suggestions of course)

Reply:
I know what pattern you're trying to implement, but it violates LSP - we can't use a local backend where we implicitly expect a remote backend: 
```
abstract class Backend {
 abstract listAssetVersions(): Promise<AssetVersions>
}

class LocalBackend extends Backend {
  /** Invalid operation. */
 override async listAssetVersions() {
  throw new Error('Invalid operation')
 }
}

class RemoteBackend extends Backend {
 override async listAssetVersions() {
  return await fetchVersions(...)
 }
}

//Somewhere in the code

function getVersions(backend: Backend) {
 return backend.listAssetVersions() // <- if we pass here a local backend, then 💥
}
```
The only solution is either to implement all methods in both the local and remote backend or not put implementation-specific methods in the parent class.

Reply:
that's intentional on our part. however, perhaps we can make those methods optional and check for existence? might not be ideal but i think feature detection like that is a decent compromise. we do want different feature sets between backends after all...

Reply:
@somebody1234 
>  we do want different feature sets between backends after all...

This is exactly why we shouldn't put methods, specific to a particular backend, into the parent superclass.

Reply:
@MrFlashAccount sure, but we don't want to have to use `instanceof` checks. what if we add another backend that *isn't* `RemoteBackend` but implements a subset (but not all) of `RemoteBackend`-specific features? do we go through all code that has `RemoteBackend` checks and add additional special-cases?

Reply:
This can be done using interfaces. We say that a function accepts anything that satisfies the interface.

Reply:
@MrFlashAccount yes, but the point is that we have a context containing a backend of an *unknown type*. in order to pass the correct type to the function, you must know the *concrete type* of the value - meaning you need to do an `instanceof` check, meaning the code needs to know about all the `Backend` subclasses again.

Reply:
And that's fine. Why would we pass a local backend if a component is supposed to work with a remote backend? We described this in types. But now a component expects any backend, and this leads to runtime errors when we pass a local backend. This means, that the local backend doesn't satisfy the `Backend` interface and shouldn't extend it.

Reply:
the entire point of having `Backend` in the first place though, is to abstract away *all* backend behavior. note that the functionality available to each backend may change in the future (see for example the recent addition of directory listing to the local backend.) - and so `instanceof RemoteBackend` simply isn't a robust solution